### ISSUE TITLE
[css-conditional-5] Feature detection for descriptors

### ISSUE THREAD

**Issue by xfq** [w3c/csswg-drafts#2463]
Currently, `@supports` and `CSS.supports` only support `property: value` pairs. The only way for web developers to detect support for a certain descriptor in an at-rule is to use JavaScript. For example:

```javascript
let e = document.createElement("style");
e.textContent = "@font-face { font-display: swap; }";
document.documentElement.appendChild(e);
let isFontDisplaySupported = e.sheet.cssRules[0].cssText.indexOf("font-display") != -1;
```

(from https://bugzilla.mozilla.org/show_bug.cgi?id=1296373#c6)

It would be good if CSS can test whether a user agent supports a `descriptor: value` pair.

See also a related issue in Chromium: https://bugs.chromium.org/p/chromium/issues/detail?id=600137

Tagged as "unknown/future spec" for now, because I'm not sure if we should extend `@supports` (which is in `css-conditional`), or develop a new mechanism for it.

**Comment by svgeesus** [comment]
I agree that being able to detect descriptor support is valuable. I would have thought it belongs in CSS Conditional.

**Comment by tabatkins** [comment]
Yeah, it's Conditional - this is precisely why we took care to include arbitrary functional terms in the @supports grammar which resolve to unknown, so we can add more types of support-testing in the future.

(Need to also be able to test for types, like if "1foo" is a `<length>`, etc.)

**Comment by tabatkins** [comment]
Something like `@supports descriptor(@media, font-display: block) { ... }`.

**Comment by frivoal** [comment]
@tabatkins Did you mean to write:
```css
@supports descriptor(@font-face, font-display: block) { ... }
```

**Comment by frivoal** [comment]
> (Need to also be able to test for types, like if "1foo" is a <length>, etc.)

Maybe. We have a clear syntax extension path if we want to, but why do we need to test if `1foo` is supported independently of testing if `bar: 1foo` is supported?

**Comment by tabatkins** [comment]
> Did you mean to write:

Yes, I did mean that. 

> Maybe. We have a clear syntax extension path if we want to, but why do we need to test if 1foo is supported independently of testing if bar: 1foo is supported?

Yeah, one could do that, but given that UAs will have explicit parsers for each value type for Typed OM purposes, we'll be able to expose that directly, if it makes one's intent clearer.

**Comment by frivoal** [comment]
Sure, but when you want to use a value, you're going to use it somewhere. And it is possible that the UA supports that value without supporting it where you planned to use it.

For instance, a newer version of a property may accept % and length while the old one only took %. The browser you're in does not yet support the length version yet, but does support the new fancy length unit you want to use in places where it does support length.

In that sense, while `@supports unit(1ch)` would something that is technically true, it seems a bit of a footgun: `@supports unit(1ch) { prop-that-used-to-take-percent-only: 1ch; }`

**Comment by ExE-Boss** [comment]
The `unit(…)` function could only be exposed from within `CSS.supports(…)` but not from `@supports`, that&nbsp;would solve the&nbsp;footgun issue while allowing developers to test for unit support from **JavaScript**.

**Comment by mrego** [comment]
It'd be nice to have this for `@property` so we could feature detect the support from CSS directly with something like:
```css
  @supports descriptor(@property, syntax: "<color>") { ... }
```

**Comment by yisibl** [comment]
> It'd be nice to have this for `@property` so we could feature detect the support from CSS directly with something like:
> 
> ```css
>   @supports descriptor(@property, syntax: "<color>") { ... }
> ```

Yeah, @property urgently needs this function, Chrome is ready to ship: https://groups.google.com/a/chromium.org/forum/#!topic/blink-dev/3ygpsew53a0

**Comment by yisibl** [comment]
Especially in [css-properties-values-api] currently implemented by Chrome, `syntax: '<image>'` is not yet supported, which makes it difficult to detect it. cc @andruud

**Comment by yisibl** [comment]
Is it possible to do a CSS @supports check on a @media rule?
https://stackoverflow.com/questions/44244221/is-it-possible-to-do-a-css-supports-check-on-a-media-rule

**Comment by andruud** [comment]
I like the `@supports descriptor()` proposal. However, I'm not sure if it covers the `@property` use-case. The `syntax` descriptor in particular accepts any `<string>` parse time, so simply checking if there's a parse error (like we do in other `@support` cases) won't be enough. Probably each descriptor would need to define how to behave when used in a `@supports descriptor()`.

I also wonder if there are any traps regarding some descriptors depending on other descriptors. For example, the following is hard to evaluate if we don't also know the `syntax`:

```
@supports descriptor(@property, initial-value: 1em) { ... }
```

For syntax `"*"` it's valid, but for e.g. `"<length>"` it's not.

**Comment by andruud** [comment]
Of course, but my point was that checking for parsing validity alone isn't enough to do what people (/ @yisibl) want for `@property`. :slightly_smiling_face:

For example `@supports descriptor(@property, syntax: "<image>") { ... }` would be `true` in Chrome, even though `<image>` syntax is not implemented.

**Comment by tabatkins** [comment]
Hm, that's true. Validity of `syntax` *does* depend on us parsing the string, so if you wrote gibberish it would be thrown out as invalid. Seems like we could similarly test for validity of non-terminals in the string? That's information known at parse-time, certainly.  I'll open an issue on the correct spec so we're not clogging up this issue. ^_^

**Comment by SebastianZ** [comment]
Just want to mention [my syntax suggestion from #5929](https://github.com/w3c/csswg-drafts/issues/5929#issuecomment-774475760) here:

```css
@supports (@page { size: 10cm; }) {
...
}
```

That would work like the existing parsing checks for properties and selectors, so no new logic here. It only requires to extend the syntax of `@supports` to take at-rules.

Sebastian

**Comment by LeaVerou** [comment]
We also need a way to query support for `@rules` (without any particular descriptor), and I would hate it if we had to include a random descriptor to do so (like the `@property` examples above). So, ideally, the syntax should allow querying for:

- Recognition of `@rule`
- Descriptor within specific `@rule`
- Descriptor:value pair within specific `@rule`

Through the same general syntax.

I like @SebastianZ's suggestion above and it does lend itself nicely to detection of `@rule` with no descriptor (`@supports (@property)` or even `@supports (@property {})`). If it's feasible parsing-wise, I think it's probably the best option.

**Comment by svgeesus** [comment]
So to rewrite the previous examples using the proposed syntax:

Checking an **at-rule**

```css
@supports (@page) {
  // woohoo we have pages
}
```

Checking a **descriptor** in a specific at-rule

```css
@supports (@font-face{ src; }) {
  // yay webfonts
}
```

Checking a **descriptor value**  in a specific at-rule

```css
@supports (@font-palette-values{ base-palette: 1; }) {
  // yay stylable color webfonts
}
```

Checking **nested at-rules** (worry about writing this grammar)


```css
@supports (@supports (@font-palette-values{ base-palette: 1; })) {
  // yay pointless recursion
}
```

**Comment by bramus** [comment]
@svgeesus At-rules such as `@keyframes` and `@scroll-timeline` require a name. Should this be included in the syntax?

E.g. 

This:

```css
@supports (@scroll-timeline name { scroll-offsets; }) {
                         /* ↑ timeline-name included */
}
```

Or that:

```css
@supports (@scroll-timeline { scroll-offsets; }) {
                         /* ↑ timeline-name not included */
}
```

**Comment by svgeesus** [comment]
@bramus good point, yes they do.

The first syntax you propose seems more natural because the second one looks like it is a descriptor. The first syntax is also closer to the syntax of the at-rule

**Comment by Loirooriol** [comment]
We can't test CSS properties without a value. So allowing to check descriptors without a value seems inconsistent.

**Comment by tabatkins** [comment]
Yeah, was coming here to bring up that same point; I don't think we need or want this syntax:

```css
@supports (@font-face{ src; }) {
  // yay webfonts
}
```

I didn't think about nested at-rules, hm. That's in a weird middle ground; it's not as trivial as "do we support this at-rule" nor is it as hands-off as "does this successfully parse". Instead you have to maintain a somewhat more complex side-list of "do we support this at-rule inside of this other at-rule", which is precisely the sort of thing that can easily get out-of-date or be missed when updating things (especially for catch-alls like @media or @supports). I suspect we might want to leave this out, and just stick with asking for a single level of support.

So in other words, I think we want to limit things to just:

1. Testing for the existence of an at-rule by name; `(@foo)` only, nothing else. (Meaning anything in the prelude is *also* left out, like `(@scroll-timeline)` with no name provided.) This is a manual test but it's *extremely* trivial and it's likely there's a list of valid at-rule names stashed in the impl anyway (I know Blink has one).
2. Testing full parsing of an at-rule, as an entire block of syntax: `(@scroll-timeline foo {})` is valid (needs a name in the prelude, but doesn't require any additional descriptors to validly parse); `(@counter-style foo { system: cyclic; symbols: a; })` is valid (needs a name in the prelude, the 'system' descriptor, and the 'symbols' or 'additive-symbols' descriptor); `(@counter-style foo {})` is false because it doesn't parse as a valid rule, due to missing its required descriptors; `(@font-face { unknown: property; })` is false due to having an unknown descriptor in it.

**Comment by Loirooriol** [comment]
> Testing full parsing of an at-rule, as an entire block of syntax

I was also thinking about this. But should this be restricted to at-rules? We could check any block I guess

```css
@supports (:foo { bar: baz; })
```

Like `@supports selector(:foo) and (bar: baz)`?

**Comment by tabatkins** [comment]
We can't do naked testing of style rules; it's the same problem as Nesting, where the grammar of properties and style rules overlap.  But yeah, testing for selectors and properties separately is fine.

**Comment by SebastianZ** [comment]
Thinking more about it, I am not feeling totally comfortable with introducing a syntax that only checks the at-rule's name. It definitely looks nicer that way but I fear it may let authors conclude that they can use the rule without limitations. Though it doesn't cover any info about its descriptors or its syntax.

Authors don't win anything if they test for an at-rule that way but the descriptors or the syntax they use isn't supported.

Sebastian

**Comment by tabatkins** [comment]
Sure, they can test for an at-rule and miss a descriptor, but the same is true of existing property tests; you can easily test for a simple version of a property and fail to realize that the more complex version you're using in your actual styles isn't supported. That's a risk authors take when they're weighing the competing concerns of terseness/readability vs accuracy, and it's not something we need to work *too* hard to protect people from, imo.

(Fun fact: an early proposal for @supports, way back when, worked by having *no* test in the prelude; instead you tagged individual properties in the contained styles with a !keyword, and it matched based on whether those properties parsed or not. No repetition at all, but what was being matched was a lot less clear up-front.)

I think the trade-off of accuracy vs terseness in allowing just the at-rule name is reasonable here. *Most* of the time, the mere presence of the at-rule is a sufficient test. When there have been gradual upgrades to an at-rule's feature set and you need to discriminate between support levels, the full-fat version of the query is there for you, but I don't think we need to force that on everyone the rest of the time. For example, I do not want people to have to write out that full `@counter-style` example just to test if the at-rule is allowed at all; that's way excessive.

(Properties, on the other hand, change their value set pretty regularly, and are a lot shorter to write, so the trade-offs of requiring the full `(prop: val)` over a simple `(prop)` isn't nearly as onerous.)

**Comment by Loirooriol** [comment]
> We can't do naked testing of style rules

Then something like `@supports block(@scroll-timeline foo {})` or `@supports block(:foo { bar: baz; })`?

It seems a bit strange to mix both properties and at-rule blocks in naked `()`, but not style blocks. May be better to just have a function for blocks in general.

**Comment by tabatkins** [comment]
Different blocks have different requirements - the child rules in @keyframes use a different prelude syntax than style rules. Can't really be generic here.

There's also the question of "why?" - what does putting an entire style rule into an @supports test achieve? If it's testing selectors, we already have `selector()`; if it's testing properties, that's the original feature. There's no interaction between the two to worry about, either.

**Comment by Loirooriol** [comment]
For consistency: if we allow at-rules but not style rules, it will probably be confusing for authors. `@keyframes` using different preludes would still need to be handled if mixing at-rules with properties in naked parenthesis, right?

**Comment by tabatkins** [comment]
> For consistency: if we allow at-rules but not style rules, it will probably be confusing for authors.

I don't think I agree? They're pretty different things. I suspect authors will be fine.

> @keyframes using different preludes would still need to be handled if mixing at-rules with properties in naked parenthesis, right?

Apologies, I'm not sure what you're asking here.

**Comment by Loirooriol** [comment]
> Apologies, I'm not sure what you're asking here.

So you said "Different blocks have different requirements - the child rules in `@keyframes` use a different prelude syntax than style rules". But this syntax needs to be supported in `@supports(@keyframes k{from{color:blue}})`. So I don't get why supporting it in a more generic function that includes style rule blocks would be problematic.

**Comment by tabatkins** [comment]
> But this syntax needs to be supported in @supports(@keyframes k{from{color:blue}})

In that example it's not any sort of special thing. You just literally feed the entire contents of the query (everything between the parens) into the parser and see if there are any parsing errors. That's just fine! No special-casing required, just possibly adding a mode to your existing parser to flag parsing errors rather than just silently dropping them.

This is very distinct from a special-case syntax for testing rule nesting.

**Comment by SebastianZ** [comment]
Quick note to everyone here: Be mindful to surround at-rules with backticks in discussions to avoid pinging people with that GitHub account name!

Sebastian

**Comment by kizu** [comment]
@tabatkins Should this issue have a “Needs Edits” label? I looked into the specs and didn't find a mention of the `at-rule()`.

Context: https://codepen.io/propjockey/pen/OJdJmya?editors=1100 — right now developers have to test for “features that co-exist” with the at-rule implementations in order to test the at-rule support. It would be awesome to have a better way to do this, and I think the sooner it would be possible to do so — the better, as more potential at-rules would get covered by this.

**Comment by stoyan** [comment]
This is much needed. Use case:

We can use `ascent-override`, `descent-override`, `line-gap-override` and `size-adjust` in `@font-face` blocks to craft a fallback for a custom font. Or at least `size-adjust` (for width) and `ascent-overrode` for height. Safari supports `size-adjust` but none of the others. Which makes the fallback worse than if it supported nothing. So I need to test for `ascent-override` support and the only robust solution is to use JavaScript like

```js
if ('ascentOverride' in new FontFace(1,1)) {...}
```

**Comment by yisibl** [comment]
@stoyan 
> Safari supports size-adjust but none of the others.

It is now supported by all major browsers. See https://web.dev/articles/css-size-adjust

**Comment by ExE-Boss** [comment]
@yisibl
> @stoyan
> 
> > Safari supports size-adjust but none of the others.
> 
> It is now supported by all major browsers. See https://web.dev/articles/css-size-adjust

**Safari 16** is still quite big, as it’s the last version to support **iPhone 8** (last **iPhone** model before **Face ID**; excluding **iPhone SE 3**) and **iPhone X**.

**Comment by stoyan** [comment]
> @stoyan
> 
> > Safari supports size-adjust but none of the others.
> 
> It is now supported by all major browsers. See https://web.dev/articles/css-size-adjust

@yisibl My point exactly. I only want a @font-face block if the browser supports `size-adjust` **and** `ascent-override`. Supporting `size-adjust` **but not** `ascent-override` makes things worse than not supporting `size-adjust` at all

**Comment by yisibl** [comment]
Microsoft's new proposal:

```css
@supports at-rule(@property; syntax: "<color>") {
    /* declare custom properties with color syntax */
}
```

See https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/AtRuleFeatureDetection/explainer.md#detect-whether-a-given-declaration-is-supported-within-an-at-rule-block

**Comment by AtkinsSJ** [comment]
See also the suggestion in https://github.com/w3c/csswg-drafts/issues/12622 which could do the same detection with something like this:

```css
@supports-condition --color-property {
	@property --foo {
		syntax: "<color>";
		inherits: false;
		initial-value: black;
	}
}
@supports --color-property {
	/* declare custom properties with color syntax */
}
```
It's definitely more verbose as afaik you'd need to provide a complete valid `@property` rule instead of just the descriptor on its own. Then again, some descriptors' requirements depend on what other descriptors are specified.
